onnxruntime/core/providers/rocm/cu_inc/common.cuh

```diff
@@ -316,7 +316,8 @@ struct alignas(sizeof(T) * vec_size) aligned_vector {
 //#define HIP_KERNEL_ASSERT(...) assert(__VA_ARGS__)
 
 // WARP related definitions and functions
-constexpr int GPU_WARP_SIZE = 64;
+constexpr int GPU_WARP_SIZE = warpSize;
+inline int GPU_WARP_SIZE_HOST= warpSizeDynamic();
```

Comment:
Are there any situations where GPU_WARP_SIZE and GPU_WARP_SIZE_HOST can be of different values? 

How do you define GPU_WARP_SIZE_HOST? Is there a way to avoid introducing this variable? It seems a confusing name to me, as warp is a concept for devices (GPUs), not for HOST (CPUs).

Reply:
Yes, GPU_WARP_SIZE is now `warpSize` which is defined in HIP headers.  It is `static constexpr int warpSize = __AMDGCN_WAVEFRONT_SIZE;`.  `__AMDGCN_WAVEFRONT_SIZE` is a hip-clang intrinsic value determined during the device compilation pass and the given gfx target value.  For MI-series gfx targets such as MI100 (gfx908) and MI200 (gfx90a), the `warpSize == 64`.  For gfx1030, `warpSize == 32`.  This can be used in `__device__` functions or `__global__` kernels to compile them with the correct `warpSize` for the given gfx arch.  If more than one gfx arch is being used for the build, each gfx target arch has its own compilation pass so this value will change as appropriate.

For host code, you cannot use `warpSize` because for the host compilation pass it defaults to 64 but it really should be undefined.  Often times `warpSize` is useful to know prior to launching kernels for setting things such as grid sizes or shared memory sizes.  We must query the device at runtime from the host code to determine the warpSize of the device.  That's why we need `warpSizeDynamic()` to do that query from host code, hence the `HOST` in the name of the new variable.

Reply:
A few questions:
(1) How do we currently build for multiple gfx archs?

(2) What does warpSizeDynamic() return? Can you give an example where warpSizeDynamic() and warpSize are of different values?

(3) Still not sure exactly why we need to introduce GPU_WARP_SIZE_HOST and we can't just use GPU_WARP_SIZE (warpSize) for all cases? Can you point out a specific example where warpSizeDynamic() is needed?

(3) Is there any reason we need to introduce GPU_WARP_SIZE_HOST for CUDA EP? I see it is simply set to GPU_WARP_SIZE.

Reply:
I tried to answer all 4 of your questions [here](https://github.com/microsoft/onnxruntime/pull/11368#discussion_r865997146).

Concerning why we need to introduce GPU_WARP_SIZE_HOST for the CUDA EP -- the ROCm EP reuses a lot of the CUDA EP source code via the source-to-source transliteration "hipify" step.  CUDA's warpSize has been 32 for all known devices up to this point, so we can safely hard-code both the GPU_WARP_SIZE and GPU_WARP_SIZE_HOST to the same 32 value.   AMD GPUs can be either warpSize 32 (gfx1030) or 64 (gfx90a) depending on the arch.

Reply:
Thanks Jeff for the detailed explanations. This looks good to me. Adding Tianlei and Weixing as reviewers to see if they have additional comments or suggestions.